freqtrade/freqai/freqai_interface.py

```diff
@@ -0,0 +1,267 @@
+        for storing, saving, loading, and analyzing the data.
+        :params:
+        :unfiltered_dataframe: Full dataframe for the current training period
+        :metadata: pair metadata from strategy.
+        :returns:
+        :model: Trained model which can be used to inference (self.predict)
+        """
+
+    @abstractmethod
+    def fit(self) -> Any:
+        """
+        Most regressors use the same function names and arguments e.g. user
+        can drop in LGBMRegressor in place of CatBoostRegressor and all data
+        management will be properly handled by Freqai.
+        :params:
+        :data_dictionary: the dictionary constructed by DataHandler to hold
+        all the training and test data/labels.
+        """
+
+        return
+
+    @abstractmethod
+    def predict(self, dataframe: DataFrame, metadata: dict) -> Tuple[npt.ArrayLike, npt.ArrayLike]:
+        """
+        Filter the prediction features data and predict with it.
+        :param: unfiltered_dataframe: Full dataframe for the current backtest period.
+        :return:
+        :predictions: np.array of predictions
+        :do_predict: np.array of 1s and 0s to indicate places where freqai needed to remove
+        data (NaNs) or felt uncertain about data (PCA and DI index)
+        """
+
+    def model_exists(self, pair: str, training_timerange: str) -> bool:
+        """
+        Given a pair and path, check if a model already exists
+        :param pair: pair e.g. BTC/USD
+        :param path: path to model
+        """
+        coin, _ = pair.split("/")
+        self.dh.model_filename = "cb_" + coin.lower() + "_" + training_timerange
+        path_to_modelfile = Path(self.dh.model_path / str(self.dh.model_filename + "_model.joblib"))
+        file_exists = path_to_modelfile.is_file()
+        if file_exists:
+            logger.info("Found model at %s", self.dh.model_path / self.dh.model_filename)
+        else:
+            logger.info("Could not find model at %s", self.dh.model_path / self.dh.model_filename)
+        return file_exists
+
+    @threaded
+    def retrain_model_on_separate_thread(self, new_trained_timerange: str, metadata: dict,
+                                         strategy: IStrategy):
+
+        # with nostdout():
+        self.dh.download_new_data_for_retraining(new_trained_timerange, metadata)
+        corr_dataframes, base_dataframes = self.dh.load_pairs_histories(new_trained_timerange,
+                                                                        metadata)
+
+        unfiltered_dataframe = self.dh.use_strategy_to_populate_indicators(strategy,
+                                                                           corr_dataframes,
+                                                                           base_dataframes,
+                                                                           metadata)
+
+        self.model = self.train(unfiltered_dataframe, metadata)
+        self.dh.save_data(self.model)
```

Comment:
You'll need to mutex a few things here i think (not necessarily on this line - but also on others where appropriate ...).

as this is running in a different thread noow, you'll have to consider that freqtrade will read the model while the new model is being written.

it's pretty easy with `from threading import Lock` - and then wrapping it in a lock instance (look at `freqtradebot.py` for an example).

Reply:
Thanks, I will put this on on the todo list.

Reply:
After further thought and reconsideration of the new code structure (I made some significant changes to how the data is handled in recent days) this is not an issue. 

Here is  why: the new model file has a different name and is being saved in a different directory with a different name. FreqAI is controlling where to look for the model file eachtime `populate_indicators()` is executed for a pair. The code is written right now so that if a new thread is spawned to train a model, FreqAI continues on to load the old model and give predictions back to the strategy. That separate thread will finish training sometime in the future and write a new file in a new directory. Next time `populate_indicators()` is called for the same pair, FreqAI directs the newest model file to be loaded.

At the moment, I dont see anywhere where we need to mutex anything, but I will keep it in mind as code develops.